### PR TITLE
JERC

### PR DESCRIPTION
**Pull request by MatthewDittrich**
In this PR:
* Implementation of the JEC and JER corrections for Run 2 and Run 3 using correctionlib
* R3 jet veto maps
* Unrelated updates also included in this PR: Also include prefixes for running with file copy at HPG 

New refs with all updates are: 4.57517, 2.49108, 5.12477

This is the version of the analysis that will be frozen for pre-approval.

### CODE REVIEW

File: analysis/wwz/get_wwz_yields.py [hunk 1]
```diff
@@ -98,7 +98,121 @@
 
 
 
-
+JERC_LST = [
```
Comment (kmohrman) [inline analysis/wwz/get_wwz_yields.py:101]: We should either not have an explicit list for each year, or we should store this long list somewhere else. 

But this is minor, let's not address it right now.

File: analysis/wwz/full_run2_run.sh [hunk 2]
```diff
@@ -16,8 +16,8 @@
 ### Examples with 4l skims configs ### (ONLY WORKS AT UAF. SAMPLES HAVE NOT BEEN MOVED TO HPG!)
 
 # Run at scale with futures
-#time python run_wwz4l.py ../../input_samples/cfgs/wwz_analysis/samples_4lskim_run2_v1.cfg -x futures -n 200 -s 100000000 -o r2_wwz_histos_noSyst --hist-list bdt
-time python run_wwz4l.py ../../input_samples/cfgs/wwz_analysis/samples_4lskim_run2_v1.cfg -x futures -n 200 -s 100000000 -o r2_wwz_histos_withSyst --hist-list njets njets_counts --do-systs
+time python run_wwz4l.py ../../input_samples/cfgs/wwz_analysis/samples_4lskim_run2_v1.cfg -x futures -n 200 -s 100000000 -o r2_wwz_histos_nosyst
+#time python run_wwz4l.py ../../input_samples/cfgs/wwz_analysis/samples_4lskim_run2_v1.cfg -x futures -n 200 -s 100000000 -o r2_wwz_histos_withSyst --hist-list njets njets_counts --do-systs
```
Comment (kmohrman) [inline analysis/wwz/full_run2_run.sh:20]: Do we need to make this change?

File: analysis/wwz/full_run3_run.sh [hunk 3]
```diff
@@ -14,13 +14,13 @@
 #################################################################################################################################
 
 #Run at scale (with futures) for JUST 2022
-#time python run_wwz4l.py ../../input_samples/cfgs/wwz_analysis/samples_4lskim_2022_v1.cfg -o y22_wwz_histos_noSyst -x futures -n 200 --hist-list bdt
```
Comment (kmohrman) [inline analysis/wwz/full_run3_run.sh:17]: Do we need these changes? Unless there is something substantial here that I'm missing, let's just stick with the current master version

File: analysis/wwz/get_wwz_yields.py [hunk 4]
```diff
@@ -519,7 +633,8 @@ def make_syst_plots(histo_dict,grouping_mc,grouping_data,save_dir_path,year):
 
     for var_name in histo_dict.keys():
         #print(f"\n{var_name}")
-        if var_name not in TMP_VAR_LST: continue
+        if var_name != "njets": continue
```
Comment (kmohrman) [inline analysis/wwz/get_wwz_yields.py:636]: Let's not change the default behavior here

File: analysis/wwz/get_wwz_yields.py [hunk 5]
```diff
@@ -546,8 +661,8 @@ def make_syst_plots(histo_dict,grouping_mc,grouping_data,save_dir_path,year):
         syst_var_lst = []
         all_syst_var_lst = histo.axes["systematic"]
         for syst_var_name in all_syst_var_lst:
-            if syst_var_name.endswith("Up"):
-                syst_name_base = syst_var_name.replace("Up","")
+            if syst_var_name.endswith("Down"):
+                syst_name_base = syst_var_name.replace("Down","")
```
Comment (kmohrman) [inline analysis/wwz/get_wwz_yields.py:665]: What's the reason for this change? If there's an up there should be a down. Why does it matter to use Down not up?

File: analysis/wwz/get_wwz_yields.py [hunk 6]
```diff
@@ -562,23 +677,39 @@ def make_syst_plots(histo_dict,grouping_mc,grouping_data,save_dir_path,year):
...
-                # Skip the variations that don't apply (TODO: why are these in the hist to begin with??)
-                if year == "UL16APV": blacklist_years = ["2016","2017","2018","2022","2022EE"]
-                if year == "UL16": blacklist_years = ["2016APV","2017","2018","2022","2022EE"]
-                if year == "UL17": blacklist_years = ["2016APV","2016","2018","2022","2022EE"]
-                if year == "UL18": blacklist_years = ["2016APV","2016","2017","2022","2022EE"]
-                if year == "2022": blacklist_years = ["2016APV","2016","2017","2018","2022EE"]
-                if year == "2022EE": blacklist_years = ["2016APV","2016","2017","2018","2022"]
-                if year == "run3": blacklist_years = ["2016APV","2016","2017","2018"]
-                if year == "all": blacklist_years = []
+                if (("uncorrelated" in syst) or (syst.startswith("JER"))):
```
Comment (kmohrman) [inline analysis/wwz/get_wwz_yields.py:689]: Can you explain conceptually what this is changing?

File: analysis/wwz/get_wwz_yields.py [hunk 7]
```diff
@@ -597,22 +728,22 @@ def make_syst_plots(histo_dict,grouping_mc,grouping_data,save_dir_path,year):
                 mc_down_arr = mc_down[{"process_grp":sum}].values()
 
                 # Print individual syst numbers
-                #if var_name != "nleps": continue
```
Comment (kmohrman) [inline analysis/wwz/get_wwz_yields.py:600]: Let's leave commented by default

File: analysis/wwz/get_wwz_yields.py [hunk 8]
```diff
@@ -950,8 +1081,8 @@ def main():
 
     # Make plots
     if args.make_plots:
-        make_plots(histo_dict,sample_dict_mc,sample_dict_data,save_dir_path=out_path,apply_nsf_to_cr=False)
-        #make_syst_plots(histo_dict,sample_dict_mc,sample_dict_data,out_path,args.ul_year) # Check on individual systematics
+        #make_plots(histo_dict,sample_dict_mc,sample_dict_data,save_dir_path=out_path,apply_nsf_to_cr=False)
```
Comment (kmohrman) [inline analysis/wwz/get_wwz_yields.py:1084]: Let's leave the normal plotting function as the uncommented one

File: ewkcoffea/data/wwz_jerc/2016APV_jerc/fatJet_jerc.json
Comment (kmohrman) [inline ewkcoffea/data/wwz_jerc/2016APV_jerc/fatJet_jerc.json:1]: For all of the files in data/wwz_jerc, are these also in topcoffea?

File: ewkcoffea/data/wwz_jerc/jer_smear.json
Comment (kmohrman) [inline ewkcoffea/data/wwz_jerc/jer_smear.json:1]: Is this straight from the POG? Is it also in topcoffea?

File: input_samples/cfgs/wwz_analysis/samples_4lskim_2022_v1.cfg [hunk 9]
```diff
@@ -19,6 +19,7 @@
 ../../input_samples/sample_jsons/wwz_analysis_4l_skims_v1/data_samples/2022EE/Muon_Run2022E-22Sep2023.json
 ../../input_samples/sample_jsons/wwz_analysis_4l_skims_v1/data_samples/2022EE/Muon_Run2022F-22Sep2023.json
 ../../input_samples/sample_jsons/wwz_analysis_4l_skims_v1/data_samples/2022EE/Muon_Run2022G-22Sep2023.json
+
```
Comment (kmohrman) [inline input_samples/cfgs/wwz_analysis/samples_4lskim_2022_v1.cfg:22]: Let's drop this so it does not show up on the diff list

File: input_samples/cfgs/wwz_analysis/samples_4lskim_run3_v1.cfg [hunk 10]
```diff
@@ -2,55 +2,55 @@
...
-../../input_samples/sample_jsons/wwz_analysis_4l_skims_v1/data_samples/2023BPix/EGamma_0_V2_Run2023D-22Sep2023.json
-../../input_samples/sample_jsons/wwz_analysis_4l_skims_v1/data_samples/2023BPix/EGamma_1_V1_Run2023D-22Sep2023.json
-../../input_samples/sample_jsons/wwz_analysis_4l_skims_v1/data_samples/2023BPix/EGamma_1_V2_Run2023D-22Sep2023.json
-../../input_samples/sample_jsons/wwz_analysis_4l_skims_v1/data_samples/2023BPix/Muon_0_V1_Run2023D-22Sep2023.json
-../../input_samples/sample_jsons/wwz_analysis_4l_skims_v1/data_samples/2023BPix/Muon_0_V2_Run2023D-22Sep2023.json
-../../input_samples/sample_jsons/wwz_analysis_4l_skims_v1/data_samples/2023BPix/Muon_1_V1_Run2023D-22Sep2023.json
-../../input_samples/sample_jsons/wwz_analysis_4l_skims_v1/data_samples/2023BPix/Muon_1_V2_Run2023D-22Sep2023.json
-../../input_samples/sample_jsons/wwz_analysis_4l_skims_v1/data_samples/2023BPix/MuonEG_V1_Run2023D-22Sep2023.json
-../../input_samples/sample_jsons/wwz_analysis_4l_skims_v1/data_samples/2023BPix/MuonEG_V2_Run2023D-22Sep2023.json
+#../../input_samples/sample_jsons/wwz_analysis_4l_skims_v1/data_samples/2022/DoubleMuon_Run2022C-22Sep2023.json
```
Comment (kmohrman) [inline input_samples/cfgs/wwz_analysis/samples_4lskim_run3_v1.cfg:5]: Was this intentional?

File: ewkcoffea/modules/corrections.py
Comment (kmohrman) [inline ewkcoffea/modules/corrections.py:1]: Do I understand correctly that is would sort of be a wrapper around the methods in topcoffea?

File: analysis/wwz/get_wwz_yields.py (hunk 1 above)
Comment (MatthewDittrich) [reply analysis/wwz/get_wwz_yields.py:101]: My initial thoughts on how to fix this is trying to get things symmetrized for each year, but this may be avoidable if we get the green light on total

File: analysis/wwz/get_wwz_yields.py (hunk 5 above)
Comment (MatthewDittrich) [reply analysis/wwz/get_wwz_yields.py:665]: Some of the JEC uncertainties end in 'Up' (PileUp), but none end in Down so far :)

File: ewkcoffea/data/wwz_jerc/2016APV_jerc/fatJet_jerc.json
Comment (MatthewDittrich) [reply ewkcoffea/data/wwz_jerc/2016APV_jerc/fatJet_jerc.json:1]: Yes they are as still zipped files. These files are not used right now, but we can probably kick this can down the road

File: ewkcoffea/data/wwz_jerc/jer_smear.json
Comment (MatthewDittrich) [reply ewkcoffea/data/wwz_jerc/jer_smear.json:1]: Straight from POG. This is the smearing correctionlib file. Topcoffea is utilizing a different smearing method (better in my opinion)

File: input_samples/cfgs/wwz_analysis/samples_4lskim_run3_v1.cfg (hunk 10 above)
Comment (MatthewDittrich) [reply input_samples/cfgs/wwz_analysis/samples_4lskim_run3_v1.cfg]: I was running tests with just MC

File: ewkcoffea/modules/corrections.py
Comment (MatthewDittrich) [reply ewkcoffea/modules/corrections.py:1]: Sort of, the tool in topcoffea is very generic (taking lists of input variables and desired corrections, uncertainties). To use that tool, you need a method to create the proper name maps and such.

File: analysis/wwz/full_run2_run.sh (hunk 2 above)
Comment (MatthewDittrich) [reply analysis/wwz/full_run2_run.sh]: fixed

File: analysis/wwz/full_run3_run.sh (hunk 3 above)
Comment (MatthewDittrich) [reply analysis/wwz/full_run3_run.sh]: fixed

File: analysis/wwz/get_wwz_yields.py (hunk 4 above)
Comment (MatthewDittrich) [reply analysis/wwz/get_wwz_yields.py]: fixed

File: analysis/wwz/get_wwz_yields.py (hunk 8 above)
Comment (MatthewDittrich) [reply analysis/wwz/get_wwz_yields.py]: fixed

File: analysis/wwz/get_wwz_yields.py (hunk 1 above)
Comment (kmohrman) [reply analysis/wwz/get_wwz_yields.py:101]: Yeah let's leave it for now, and revisit once we get a GL on Total (hopefully)

File: ewkcoffea/data/wwz_jerc/2016APV_jerc/fatJet_jerc.json
Comment (kmohrman) [reply ewkcoffea/data/wwz_jerc/2016APV_jerc/fatJet_jerc.json:1]: See https://github.com/cmstas/ewkcoffea/issues/54

File: ewkcoffea/data/wwz_jerc/jer_smear.json
Comment (kmohrman) [reply ewkcoffea/data/wwz_jerc/jer_smear.json:1]: See https://github.com/cmstas/ewkcoffea/issues/54

File: ewkcoffea/modules/corrections.py
Comment (kmohrman) [reply ewkcoffea/modules/corrections.py:1]: Then this should maybe also live in topcoffea? Unless it's analysis specific I guess. Anyway we can figure it out later

File: analysis/wwz/wwz4l.py [hunk 11]
```diff
@@ -452,9 +477,31 @@ def process(self, events):
...
+            cleanedJets["pt_raw"] = (1 - cleanedJets.rawFactor)*cleanedJets.pt
+            cleanedJets["mass_raw"] = (1 - cleanedJets.rawFactor)*cleanedJets.mass
+
+            if not isData:
+                cleanedJets["pt_gen"] =ak.values_astype(ak.fill_none(cleanedJets.matched_gen.pt, 0), np.float32)
+            else:
+                cleanedJets["pt_gen"] =ak.ones_like(cleanedJets.pt)
+
+            if year in ["2022","2022EE","2023","2023BPix"]:
+                cleanedJets["rho"] = ak.broadcast_arrays(events.Rho.fixedGridRhoFastjetAll, cleanedJets.pt)[0]
```
Comment (kmohrman) [inline analysis/wwz/wwz4l.py:493]: What does this broadcast_arrays do, can you add a comment to explain? Also, can you just use the `rho` that you defined at the top?

File: analysis/wwz/wwz4l.py [hunk 12]
```diff
@@ -452,9 +477,31 @@ def process(self, events):
...
+            elif year in ["2016APV","2016","2017","2018"]:
+                cleanedJets["rho"] = ak.broadcast_arrays(events.fixedGridRhoFastjetAll, cleanedJets.pt)[0]
+
```
Comment (kmohrman) [inline analysis/wwz/wwz4l.py:496]: Can you add a comment to the code about this events_cache thing? I.e. just like one line saying something about what it is or what it does

File: analysis/wwz/wwz4l.py [hunk 13]
```diff
@@ -584,7 +631,10 @@ def process(self, events):
             ######### Masks we need for the selection ##########
 
             # Pass trigger mask
```
Comment (kmohrman) [inline analysis/wwz/wwz4l.py:633]: If I understand properly, the era is not wanted for this function for R2, but it's not None. Instead of putting the whole call in an if statement (which is hard to read), I think it would be more clear to just use a statement to set a bool and then only have one function call as before. I.e. add this:
```
era_for_trg_check = era
if not (is2022 or is2023):
    era_for_trg_check = None
```
Then just call the `es_tc.trg_pass_no_overlap()` once and pass `era_for_trg_check` to it for era.

File: analysis/wwz/get_wwz_yields.py (hunk 6 above)
Comment (MatthewDittrich) [reply analysis/wwz/get_wwz_yields.py]: no clue, and I don't see why it is needed. I have removed it for now, but can always add this back in if for some reason it is helpful

File: analysis/wwz/wwz4l.py (hunk 11 above)
Comment (MatthewDittrich) [reply analysis/wwz/wwz4l.py]: added comment, but will explain here as well:

this broadcasts the rho variable to have the same shape as correctedJets

File: analysis/wwz/wwz4l.py (hunk 12 above)
Comment (MatthewDittrich) [reply analysis/wwz/wwz4l.py:502]: events_cache is used to store intermediate corrections calculations (as the JEC corrections are a step-by-step process)

File: analysis/wwz/wwz4l.py (hunk 13 above)
Comment (MatthewDittrich) [reply analysis/wwz/wwz4l.py:639]: done

File: analysis/wwz/wwz4l.py (hunk 11 above)
Comment (kmohrman) [reply analysis/wwz/wwz4l.py]: Thanks for the comment. 

But for the second part of the comment, can we not just use the `rho` you defined at the top of the processor? Why do you need to do events.Rho.etc here too?

File: analysis/wwz/wwz4l.py [hunk 14]
```diff
@@ -240,12 +240,12 @@ def process(self, events):
         else:
             raise Exception(f"ERROR: Unknown year {year}.")
 
-        # If this is a 2022 sample, get the era info
```
Comment (kmohrman) [inline analysis/wwz/wwz4l.py:243]: This is relatively compact, but sorry it makes my brain really hurt. Two equals lines in one line? Did not even know you could do that. 

To keep things simpler so we don't have to think while reading the code, I suggest we leave this part of the code alone (just leave it how it was before with setting eras). 

Then right above the trigger check, just add this:
```
era_for_trg_check = era
if not (is2022 or is2023):
    era_for_trg_check = None
```

File: analysis/wwz/wwz4l.py (hunk 13 above)
Comment (kmohrman) [reply analysis/wwz/wwz4l.py:639]: See comment above where you changed how we define eras. I think it's a lot easier to read if we leave that alone and just directly add this `era_for_trg_check` statement here directly above the trigger check line.

File: analysis/wwz/wwz4l.py (hunk 11 above)
Comment (kmohrman) [reply analysis/wwz/wwz4l.py]: Oh you did replace with `rho`. Ok, so now that the if and elif do exactly the same thing here, we can get rid of if/elif and just do the thing

**Comment by kmohrman** [comment]
For reference, this is the yield difference we have for R2 for applying the JERC corrections: 
Screenshot 2024-10-06 at 6 41 25 PM

**Comment by kmohrman** [comment]
For reference, this is the yield difference we have for R3 for applying the JERC corrections:
Screenshot 2024-10-06 at 6 45 08 PM

**Comment by kmohrman** [comment]
For reference, this is the yield difference we have for R3 for applying the Jet veto maps on top of the JERC corrections (with respect to just JERC corrections): 
Screenshot 2024-10-06 at 6 48 36 PM

File: analysis/wwz/wwz4l.py (hunk 13 above)
Comment (kmohrman) [reply analysis/wwz/wwz4l.py:639]: done
